vars/kaGit.groovy

```diff
@@ -43,48 +43,114 @@ def checkoutJenkinsTools() {
+   return lsRemoteLine.split("\t")[0].trim();
+}
+
+// Return each ref that was output by ls-remote for a given commit-ish.
+// ls-remote does not return any results when commit hashes are queried
+// directly.
+//
+// Example output: 
+// $ git ls-remote origin foo
+// 0670bc5a1c0bab364dfb981f7854b6b17bdd49db  refs/heads/deploy/foo
+// df49834270ad034ecf776590908d429a8140c485  refs/heads/foo
+// 2998fe06daa988757b847afd5a99022334b90d4d  refs/tags/foo
+Ref[] lsRemote(String repo, String committish) {
+   // https://git-scm.com/docs/git-ls-remote
+   String lsRemoteOutput = exec.outputOf(["git", "ls-remote", "-q", repo,
+                                          committish]);
+   if (lsRemoteOutput == "") {
+      return [];
    }
-   if (sha1) {
-      echo("'${commit}' resolves to ${sha1}");
-      return sha1;
+   
+   return lsRemoteOutput.split("\n").collect {
+      new Ref(hash: _commitHashFromLsRemoteLine(it),
+              name: _refNameFromLsRemoteLine(it));
+   };
+}
+
+// Return commit hash from a ls-remote result if any refs names are an exact
+// match to the branch name. A partial match is not performed here because we
+// don't want to return the hash for `deploy/foo` when the user requested just
+// `foo`.
+String _findCommitHashFromBranchRef(Ref[] refs, String branchName) {
+   Ref found = refs.find { it.name == "refs/heads/${branchName}" }
+   if (found) {
+      return found.hash;
+   }
+   return null;
+}
+
+// Return commit hash from a ls-remote result if any ref names are an exact
+// match to the tag name. A partial match is not performed here because we don't
+// want to return the hash for `deploy/foo` when the user requested just `foo`.
+String _findCommitHashFromTagRef(Ref[] refs, String tagName) {
+   Ref found = refs.find { it.name == "refs/tags/${tagName}" }
+   if (found) {
+      return found.hash;
    }
-   // If this looks like a sha1 already, return it.
-   // TODO(csilvers): complain to slack?
-   if (commit ==~ /[0-9a-fA-F]{5,}/) {
-      return commit;
+   return null;
+}
+
+// Return a commit ID (sha1 hash) from a commit-ish. If a branch or tag name, we
+// check if it exists on the remote and get the hash from there, preferring
+// branches over tags. Otherwise, if the input looks like a hash we return it
+// verbatim. Otherwise, we error.
+//
+// To be called before checking out the repo to determine which commit to 
+// checkout.
+// https://git-scm.com/docs/gitglossary#Documentation/gitglossary.txt-commit-ishalsocommittish
+String resolveCommitish(String repo, String committish) {
+   String hash = null
+   stage("Resolving commit-ish") {
+      timeout(time: 1, unit: "HOURS") {
+         Ref[] lsRemoteRefs = lsRemote(repo, committish);
+
+         // First, check for branch as we prefer branches over tags.
+         hash = _findCommitHashFromBranchRef(lsRemoteRefs, committish);
+         if (hash) {
+            echo("'${committish}' is a branch that resolves to ${hash}");
+            return;
```

Comment:
I don't know groovy well, what is this `return` returning from?  Clearly not the top-level function (though I don't know why).

Can you say "return hash;" here?  It would definitely read a lot better, but maybe give a false sense of what the code is doing if it's not actually returning from the top level function.

Reply:
That's what I tried first! But a return from inside a `stage() {}` just exits that stage early then continues executing the rest of the function.

We need the `stage() {}` here to be able to use `timeout() {}`. Maybe it would be less confusing if I wrapped resolveCommitish in a timeout everywhere it's used instead of the timeout being part of the shared function?

Reply:
Or the alternative: move the timeout (and the stage) to the caller.

Reply:
Err, sorry I got some of the behavior mixed up. Even wrapping in `timeout() {}` or `withTimeout() {}` results in capturing the return. I'll move the timeouts to the callers.